### PR TITLE
avoid global buffers

### PR DESCRIPTION
**Pull request by hannesm**
this should fix most of #186 -- performance-wise (tested on fortuna), this gives some speedup (but obviously should be tested a bit more)

//cc @reynir @dinosaure

### CODE REVIEW

File: rng/entropy.ml [hunk 1]
```diff
@@ -121,12 +121,11 @@ let bootstrap id =
   | Ok cpu_rng_bootstrap ->
     try cpu_rng_bootstrap id with Failure _ -> whirlwind_bootstrap id
 
-let interrupt_hook () =
+let interrupt_hook () () =
```
Comment (reynir) [inline rng/entropy.ml:124]: We could lose one of the arguments now, but it would change the interface and be a breaking change. Probably not worth it.

**Review by reynir — APPROVED** [review]
On my system the performance seems to be about the same:

### aes-128-gcm

| bytes | main | d3e0a5785c5ec76713867b715ae6af1647ebeef2 | 26e44814e661dfaf8d14a0bfb29f28680a7dc68e |
| - | - | - | - |
| 16 | 107.919318 | 108.127357 | 102.645635 |
| 64 | 352.568878 | 348.686773 | 336.847306 |
| 256 | 1127.308329 | 1126.252432 | 1074.113642 |
| 1024 | 1991.705995 | 1976.265390 | 1913.386337 |
| 8192 | 1385.496016 | 1380.143187 | 1404.725130 |

### fortuna

| bytes | main | d3e0a5785c5ec76713867b715ae6af1647ebeef2 | 26e44814e661dfaf8d14a0bfb29f28680a7dc68e |
| - | - | - | - |
| 16 | 102.238773 | 102.721865 | 99.792642 |
| 64 | 369.019572 | 366.661943 | 354.053437 |
| 256 | 1205.616092 | 1200.784292 | 1163.625689 |
| 1024 | 2440.599733 | 2447.444005 | 2412.249927 |
| 8192 | 2258.243312 | 2233.075545 | 2114.863257 |

File: rng/entropy.ml (hunk 1 above)
Comment (hannesm) [reply rng/entropy.ml]: well spotted, removed in 26e4481

**Comment by hannesm** [comment]
> On my system the performance seems to be about the same:
> ### aes-128-gcm
> bytes 	main 	[d3e0a57](https://github.com/mirage/mirage-crypto/commit/d3e0a5785c5ec76713867b715ae6af1647ebeef2)

nice to see these numbers. I guess you'll need to rerun with [0cab4b0](https://github.com/mirage/mirage-crypto/pull/219/commits/0cab4b0649bb7686f612a4c82132d2483c7a3100) since there were still some `Bytes.make` global buffers.

**Comment by hannesm** [comment]
my benchmarks:

## aes-128-gcm

| bytes | | v0.11.3 | 2e67decd | 26e4481 |
|-|-|-|-|
|   16 |  29.240972 |  56.823077 | 65.456465 |
|   64 |  95.538922 |  190.837645 | 212.013775 |
|  256 |  364.810213 | 626.125396 | 696.529714 |
| 1024 |  768.352643 | 1167.826327 | 1294.554864 |
| 8192 |  1494.688871 | 1044.287684 | 1166.427176 |

## fortuna

| bytes | | v0.11.3 | 2e67decd | 26e4481 |
|-|-|-|-|
|   16 |  40.128478 |  61.275627 | 69.199389 |
|   64 |  145.876685 | 201.986627 | 243.259022 |
|  256 |  505.176774 | 697.574029 | 821.534291 |
| 1024 |  1257.045894 | 1439.696009 | 1657.616035 |
| 8192 |  2201.152848 | 1473.062793 | 1691.881670 |

**Comment by reynir** [comment]
I updated the table. It seems slightly slower on my system now. I'm not sure it's significant.

**Comment by hannesm** [comment]
thanks @reynir. if you run it 5 times, do you reproduce these slower numbers, or do they vary a lot?

**Comment by reynir** [comment]
I ran it a few times. It seems fairly consistent that for short inputs it's slightly slower. For longer inputs it's about the same or in the case of aes-128-gcm maybe slightly faster.

**Comment by hannesm** [comment]
..slightly slower -- compared to the ~33% performance decrease measured in #186, the slight decrease observed here is IMHO fine. of course other opinions are welcome (and also more measurements).

in the meantime, I'll merge - we can iterate on this code before release.

**Pull request merged**